fuel-core/src/schema/contract.rs

```diff
@@ -113,4 +121,112 @@ impl ContractBalanceQuery {
             asset_id,
         })
     }
+
+    async fn contract_balances(
+        &self,
+        ctx: &Context<'_>,
+        filter: ContractBalanceFilterInput,
+        first: Option<i32>,
+        after: Option<String>,
+        last: Option<i32>,
+        before: Option<String>,
+    ) -> async_graphql::Result<Connection<AssetId, ContractBalance, EmptyFields, EmptyFields>> {
+        let db = ctx.data_unchecked::<Database>().clone();
+
+        query(
+            after,
+            before,
+            first,
+            last,
+            |after: Option<AssetId>, before: Option<AssetId>, first, last| async move {
+                let (records_to_fetch, direction) = if let Some(first) = first {
+                    (first, IterDirection::Forward)
+                } else if let Some(last) = last {
+                    (last, IterDirection::Reverse)
+                } else {
+                    (0, IterDirection::Forward)
+                };
+
+                let after = after.map(fuel_tx::AssetId::from);
+                let before = before.map(fuel_tx::AssetId::from);
+
+                let start;
+                let end;
+
+                if direction == IterDirection::Forward {
+                    start = after;
+                    end = before;
+                } else {
+                    start = before;
+                    end = after;
+                }
+                // impl Iterator<Item = ContractBalance>
+                let mut balances = db
+                    .contract_balances(filter.contract.into(), start, Some(IterDirection::Forward))
+                    .map(|balance| -> ContractBalance {
+                        let asset_id = balance.unwrap();
+
+                        let result =
+                            fuel_vm::storage::InterpreterStorage::merkle_contract_asset_id_balance(
+                                &db,
+                                &filter.contract.into(),
+                                &asset_id,
+                            )
+                            .unwrap();
+
+                        let amount = result.unwrap_or(0);
+
+                        ContractBalance {
+                            contract: filter.contract.into(),
+                            amount,
+                            asset_id,
+                        }
+                    })
```

Comment:
This call to collect will load every balance after the starting asset id. We need to limit how many items the iterator will possibly consume:

```suggestion
                    })
                    .take(records_to_fetch)
```